src/kernel/syscalls.zig

```diff
@@ -97,6 +181,122 @@ pub fn handle(syscall: Syscall, arg1: usize, arg2: usize, arg3: usize, arg4: usi
+///
+/// Get a slice containing the data at an address and length. If the current task is a kernel task then a simple pointer to slice conversion is performed,
+/// otherwise the slice is allocated on the heap and the data is copied in from user space.
+///
+/// Arguments:
+///     IN ptr: usize - The slice's address
+///     IN len: usize - The number of bytes
+///
+/// Error: Error
+///     Error.OutOfMemory - There wasn't enough kernel (heap or VMM) memory left to fulfill the request.
+///     Error.TooBig - The user task requested to have too much data copied
+///     Error.InvalidAddress - The address that the user task passed is invalid (not mapped, out of bounds etc.)
+///
+/// Return: []u8
+///     The slice of data. Will be stack-allocated if the current task is kernel-level, otherwise will be heap-allocated
+///
+fn getData(ptr: usize, len: usize) Error![]u8 {
+    if (scheduler.current_task.kernel) {
+        return @intToPtr([*]u8, ptr)[0..len];
+    } else {
+        if (len > USER_MAX_DATA_LEN) {
+            return Error.TooBig;
+        }
+        var buff = try allocator.alloc(u8, len);
+        errdefer allocator.free(buff);
+        try vmm.kernel_vmm.copyData(scheduler.current_task.vmm, buff, ptr, false);
+        return buff;
+    }
+}
+
+fn handleOpen(path_ptr: usize, path_len: usize, flags: usize, args: usize, ignored: usize) Error!usize {
+    const current_task = scheduler.current_task;
+    if (!current_task.hasFreeVFSHandle()) {
+        return Error.NoMoreFSHandles;
+    }
+
+    // Fetch the open arguments from user/kernel memory
+    var open_args: vfs.OpenArgs = if (args == 0) .{} else blk: {
+        const data = try getData(args, @sizeOf(vfs.OpenArgs));
+        defer if (!current_task.kernel) allocator.free(data);
+        break :blk std.mem.bytesAsValue(vfs.OpenArgs, data[0..@sizeOf(vfs.OpenArgs)]).*;
+    };
+    // The symlink target could refer to a location in user memory so convert that too
+    if (open_args.symlink_target) |target| {
+        open_args.symlink_target = try getData(@ptrToInt(target.ptr), target.len);
+    }
+    defer if (!current_task.kernel) if (open_args.symlink_target) |target| allocator.free(target);
+
+    const open_flags = std.meta.intToEnum(vfs.OpenFlags, flags) catch return Error.InvalidFlags;
+    const path = try getData(path_ptr, path_len);
+    defer if (!current_task.kernel) allocator.free(path);
+
+    var node = try vfs.open(path, true, open_flags, open_args);
+    return current_task.addVFSHandle(node) orelse panic(null, "Failed to add a VFS handle to current_task\n", .{});
+}
+
+fn handleRead(node_handle: usize, buff_ptr: usize, buff_len: usize, ignored1: usize, ignored2: usize) Error!usize {
+    if (buff_len > USER_MAX_DATA_LEN) {
+        return Error.TooBig;
+    }
+
+    const current_task = scheduler.current_task;
+    const node_opt = current_task.getVFSHandle(node_handle) catch panic(@errorReturnTrace(), "Failed to get VFS node for handle {}\n", .{node_handle});
```

Comment:
This erroring wouldn't be the fault of the caller though.

Reply:
But the user will pass the node_handle?

Reply:
Yes but `getVFSHandle` won't fail if the handle doesn't exist as it will then return null. It will only fail if something goes wrong with the map itself.